File: src/main/clj/fpinclojure/exercises/datastructures/tree.clj
(ns fpinclojure.exercises.datastructures.tree
  (:refer-clojure :exclude [map])
  (:require
   [clojure.spec.alpha :as s]))

(defrecord Leaf [value])
(defrecord Branch [left right])

(s/fdef leaf?
  :args (s/cat :x any?)
  :ret boolean?)

(defn leaf? [x]
  (instance? Leaf x))

(s/fdef branch?
  :args (s/cat :x any?)
  :ret boolean?)

(defn branch? [x]
  (and (instance? Branch x)
       ((some-fn leaf? branch?) (:left x))
       ((some-fn leaf? branch?) (:right x))))

(s/def ::tree
  (s/or :leaf leaf?
        :branch branch?))

(s/fdef size
  :args (s/cat :t ::tree)
  :ret int?)

(defn size [t]
  (if (leaf? t)
    1
    (+ (-> t :left size)
       1
       (-> t :right size))))

;; Exercise 3.25

(s/fdef maximum
  :args (s/cat :t ::tree)
  :ret any?)

(defn maximum [t]
  (if (leaf? t)
    (:value t)
    (max (-> t :left maximum)
         (-> t :right maximum))))

;; Exercise 3.26

(s/fdef depth
  :args (s/cat :t ::tree)
  :ret int?)

(defn depth [t]
  (if (leaf? t)
    0
    (inc (max (-> t :left depth)
              (-> t :right depth)))))

;; Exercise 3.27

(s/fdef map
  :args (s/cat :f ifn?
               :t ::tree)
  :ret ::tree)

(defn map [f t]
  (if (leaf? t)
    (-> t :value f ->Leaf)
    (->Branch (map f (:left t))
              (map f (:right t)))))

;; Exercise 3.28

(s/fdef fold
  :args (s/cat :f ifn?
               :g ifn?
               :t ::tree)
  :ret any?)

(defn fold [f g t]
  (if (leaf? t)
    (-> t :value f)
    (g (fold f g (:left t))
       (fold f g (:right t)))))

(s/fdef size-via-fold
  :args (s/cat :t ::tree)
  :ret int?)

(defn size-via-fold [t]
  (fold (constantly 1) #(+ %1 1 %2) t))

(s/fdef maximum-via-fold
  :args (s/cat :t ::tree)
  :ret any?)

(defn maximum-via-fold [t]
  (fold identity max t))

(s/fdef depth-via-fold
  :args (s/cat :t ::tree)
  :ret int?)

(defn depth-via-fold [t]
  (fold (constantly 0) #(inc (max %1 %2)) t))

(s/fdef map-via-fold
  :args (s/cat :f ifn?
               :t ::tree)
  :ret ::tree)

(defn map-via-fold [f t]
  (fold #(-> % f ->Leaf) ->Branch t))

(comment
  (require '[clojure.spec.test.alpha :as stest])
  (stest/instrument)

  (size (->Leaf 2))
  (size (->Branch (->Leaf 2)
                  (->Leaf 4)))
  (size (->Branch (->Leaf 1)
                  (->Branch (->Leaf 4)
                            (->Leaf 2))))

  (maximum (->Leaf 2))
  (maximum (->Branch (->Leaf 2)
                     (->Leaf 4)))
  (maximum (->Branch (->Leaf 1)
                     (->Branch (->Leaf 4)
                               (->Leaf 2))))

  (depth (->Leaf 2))
  (depth (->Branch (->Leaf 2)
                   (->Leaf 4)))
  (depth (->Branch (->Leaf 1)
                   (->Branch (->Leaf 4)
                             (->Leaf 2))))

  (map #(* % %) (->Leaf 2))
  (map #(* % %) (->Branch (->Leaf 2)
                          (->Leaf 4)))
  (map #(* % %) (->Branch (->Leaf 1)
                          (->Branch (->Leaf 4)
                                    (->Leaf 2))))

  (size-via-fold (->Leaf 2))
  (size-via-fold (->Branch (->Leaf 2)
                           (->Leaf 4)))
  (size-via-fold (->Branch (->Leaf 1)
                           (->Branch (->Leaf 4)
                                     (->Leaf 2))))
  (maximum-via-fold (->Leaf 2))
  (maximum-via-fold (->Branch (->Leaf 2)
                              (->Leaf 4)))
  (maximum-via-fold (->Branch (->Leaf 1)
                              (->Branch (->Leaf 4)
                                        (->Leaf 2))))
  (depth-via-fold (->Leaf 2))
  (depth-via-fold (->Branch (->Leaf 2)
                            (->Leaf 4)))
  (depth-via-fold (->Branch (->Leaf 1)
                            (->Branch (->Leaf 4)
                                      (->Leaf 2))))
  (map-via-fold #(* % %) (->Leaf 2))
  (map-via-fold #(* % %) (->Branch (->Leaf 2)
                                   (->Leaf 4)))
  (map-via-fold #(* % %) (->Branch (->Leaf 1)
                                   (->Branch (->Leaf 4)
                                             (->Leaf 2))))
  )

File: src/main/scala/fpinscala/exercises/datastructures/Tree.scala
package fpinscala.exercises.datastructures

enum Tree[+A]:
  case Leaf(value: A)
  case Branch(left: Tree[A], right: Tree[A])

  def size: Int = this match
    case Leaf(_) => 1
    case Branch(l, r) => 1 + l.size + r.size

  // Exercise 3.26
  def depth: Int = this match
    case Leaf(_)      => 0
    case Branch(l, r) => 1 + (l.depth max r.depth)

  // Exercise 3.27
  def map[B](f: A => B): Tree[B] = this match
    case Leaf(v)      => Leaf(f(v))
    case Branch(l, r) => Branch(l.map(f), r.map(f))

  // Exercise 3.28

  def fold[B](f: A => B, g: (B,B) => B): B = this match
    case Leaf(v)      => f(v)
    case Branch(l, r) => g(l.fold(f, g), r.fold(f, g))

  def sizeViaFold: Int =
    fold(_ => 1, _ + 1 + _)

  def depthViaFold: Int =
    fold(_ => 0, (d1, d2) => 1 + (d1 max d2))

  def mapViaFold[B](f: A => B): Tree[B] =
    fold(v => Leaf(f(v)), Branch(_, _))

object Tree:

  def size[A](t: Tree[A]): Int = t match
    case Leaf(_) => 1
    case Branch(l,r) => 1 + size(l) + size(r)

  extension (t: Tree[Int]) def firstPositive: Int = t match
    case Leaf(i) => i
    case Branch(l, r) =>
      val lpos = l.firstPositive
      if lpos > 0 then lpos else r.firstPositive

  // Exercise 3.25
  extension (t: Tree[Int]) def maximum: Int = t match
    case Leaf(v)      => v
    case Branch(l, r) => l.maximum max r.maximum

  // Exercise 3.28
  extension (t: Tree[Int]) def maximumViaFold: Int =
    t.fold(identity, _ max _)
